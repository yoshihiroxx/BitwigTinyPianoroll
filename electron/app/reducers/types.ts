import { Dispatch as ReduxDispatch, Store as ReduxStore, Action } from 'redux';
import { BrowserWindow } from 'electron';
import { PreferencesRecordType } from '../models/Preferences';
import { EditorRecordType } from '../models/Editor';

export type counterStateType = {
  counter: number;
};

export type ModelType = {
  preferences: PreferencesRecordType;
  editor: EditorRecordType;
};

export type GetState = () => ModelType;

export type Dispatch = ReduxDispatch<Action<string>>;

export type Store = ReduxStore<ModelType, Action<string>>;

export type ActionType = {
  type: string;
  payload: any;
  meta: any;
  error: boolean;
};

export type AppWindowsType = {
  [index: string]: BrowserWindow;
};
