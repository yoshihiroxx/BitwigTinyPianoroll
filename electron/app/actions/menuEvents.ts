export const LOAD_MIDIFILE = 'LOAD_MIDIFILE';
export const NEW_MIDIFILE = 'NEW_MIDIFILE';

export function onLoadMidiFile(parsedMidi: unknown) {
  return {
    type: LOAD_MIDIFILE,
    payload: {
      parsedMidi
    },
    error: null
  };
}

export function onCreateNewMidiFile() {
  return {
    type: NEW_MIDIFILE,
    payload: {},
    error: null
  };
}
