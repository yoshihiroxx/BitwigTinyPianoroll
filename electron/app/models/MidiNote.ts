import { Record, List } from 'immutable';

const MidiNoteRecord = Record({
  noteNumber: 0,
  startBeat: 0,
  lengthInBeats: 0.25,
  velocity: 60
});

export default class MidiNote extends MidiNoteRecord {
  setVelocity(velocity: number) {
    return this.set('startBeat', velocity);
  }
}
