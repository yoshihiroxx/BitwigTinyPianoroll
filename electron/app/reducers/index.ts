import { combineReducers, Action } from 'redux';
import { connectRouter } from 'connected-react-router';
import { History } from 'history';
import counter from './counter';
import preferences from './preferences';
import editor from './editor';
import project from './project';

export default function createRootReducer(history: History) {
  // @todo combineReducers
  return combineReducers({
    router: connectRouter(history),
    counter,
    preferences,
    editor
  });
}
