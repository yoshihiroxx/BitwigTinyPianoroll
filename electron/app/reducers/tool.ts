import { Action } from 'redux';
import { ActionType } from './types';
import {
  PenTool,
  EraserTool,
  MoveTool,
  RectTool,
  LengthTool
} from '../tool/Tools';
import {
  HANDLE_MOUSE_EVENT,
  HANDLE_TOOL,
  SET_SELECTIONS,
  handleMouseEvent,
  handleTool,
  focusClip,
  SET_KEEP_SELECTIONS
} from '../actions/tool';
import MidiNote from '../models/MidiNote';
import MidiList from '../models/MidiList';

type Actions =
  | ReturnType<typeof handleMouseEvent>
  | ReturnType<typeof handleTool>;

export default function tool(state = new PenTool(), action: Actions) {
  switch (action.type) {
    case HANDLE_MOUSE_EVENT: {
      switch (action.meta.event) {
        case 'click': {
          if (typeof action.payload.beatOrNote === 'number') {
            return state.onClick(
              action.payload.beatOrNote,
              action.payload.noteNumber
            );
          }
          if (action.payload.beatOrNote instanceof MidiNote) {
            return state.onClick(action.payload.beatOrNote);
          }
          return state;
        }
        case 'drag':
          return state.onDrag(
            action.payload.beatOrNote,
            action.payload.noteNumber
          );
        case 'release': {
          const nextState = state.onRelease(
            action.payload.beatOrNote,
            action.payload.noteNumber
          );
          return nextState;
        }
        default:
          return state;
      }
    }
    case HANDLE_TOOL: {
      // let nextState = state;
      // if (action.meta.force) {
      //   nextState = nextState.set('isDrawing', false);
      //   nextState = nextState.set('drawing', new MidiList());
      // }
      if (state.get('isDrawing')) return state;
      switch (action.meta.toolType) {
        case 'pen':
          return new PenTool(state.prepareToChange());

        case 'eraser':
          return new EraserTool(state.prepareToChange());

        case 'move':
          return new MoveTool(state.prepareToChange());

        case 'rect':
          return new RectTool(state.prepareToChange());

        case 'length':
          return new LengthTool(state.prepareToChange());

        default:
          throw new Error(`${action.meta.toolType}: This tool is not defined.`);
          return state;
      }
    }
    case SET_SELECTIONS: {
      return state.set('selections', action.payload.selections);
    }

    case SET_KEEP_SELECTIONS: {
      return state.set('shouldKeepSelection', action.payload.value);
    }

    default:
      return state;
  }
}
