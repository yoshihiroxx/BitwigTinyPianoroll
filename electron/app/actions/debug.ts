import { remote } from 'electron';
import { GetStage, Dispatch } from '../reducers/types';

export const LOAD_MIDIFILE = 'LOAD_MIDIFILE';

export function loadMidiFile(filename: string) {
  return (dispatch: Dispatch, getState: GetState) => {
    const state = getState();
    console.log(state);
    const baseDir = `/Users/y/work/BitwigTinyPianoroll/electron/test/midi/${filename}`;
    const readfile = path => {
      return new Promise((resolve, reject) => {
        remote.require('fs').readFile(path, 'base64', (err, data) => {
          return err ? reject(err) : resolve(data);
        });
      });
    };
    readfile(baseDir)
      .then(function(data) {
        console.log(`The file content is : ${data}`);
        const payload = {
          data
        };
        dispatch({
          type: LOAD_MIDIFILE,
          payload
        });
        return true;
      })
      .catch(function(error) {
        console.log(error);
        return false;
      });
  };
}
