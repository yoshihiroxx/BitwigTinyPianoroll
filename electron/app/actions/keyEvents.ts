import KeyBinds, { KeyBind } from '../models/KeyBinds';
import {
  setSelections,
  addNote,
  removeNote,
  clearSelections,
  setKeepSelections
} from './tool';
import { GetState, Dispatch } from '../reducers/types';

function matchedActionName(event, keyBinds: KeyBinds, isKeyUp: boolean) {
  let matchedAction = '';
  const actionNames: Array<string> = [
    'copy',
    'paste',
    'pitchUp',
    'pitchDown',
    'increaseBeat',
    'decreaseBeat',
    'removeNotes',
    'keepSelectionNotes'
  ];
  actionNames.forEach((actionName: string) => {
    const keyBind: KeyBind = keyBinds.get(actionName);
    if (!keyBind) {
      throw new Error(`${actionName}: is not defined in KeyBinds`);
    }
    if (
      keyBind.codes.some((value: number) => {
        return value === event.keyCode;
      }) &&
      ((keyBind.metaKey === event.metaKey &&
        keyBind.altKey === event.altKey &&
        keyBind.shiftKey === event.shiftKey) ||
        isKeyUp)
    ) {
      matchedAction = actionName;
    }
  });
  return matchedAction;
}

export function onKeyDown(event) {
  return (dispatch: Dispatch, getState: GetState) => {
    const { editor, preferences } = getState();
    const tool = editor.get('tool');

    const matchedAction = matchedActionName(event, preferences.keyBinds, false);

    switch (matchedAction) {
      case 'pitchUp':
        {
          const selections = tool.get('selections');
          if (!tool.get('isDrawing')) {
            selections.get('notes').forEach((note: MidiNote) => {
              dispatch(removeNote(note));
            });
            selections.get('notes').forEach((note: MidiNote) => {
              dispatch(
                addNote(note.set('noteNumber', note.get('noteNumber') + 1))
              );
            });
            dispatch(setSelections(selections.slidePitch(1)));
          }
        }
        break;
      case 'pitchDown':
        {
          const selections = tool.get('selections');
          if (!tool.get('isDrawing')) {
            selections.get('notes').forEach((note: MidiNote) => {
              dispatch(removeNote(note));
            });
            selections.get('notes').forEach((note: MidiNote) => {
              dispatch(
                addNote(note.set('noteNumber', note.get('noteNumber') - 1))
              );
            });
            dispatch(setSelections(selections.slidePitch(-1)));
          }
        }
        break;
      case 'increaseBeat':
        {
          const selections = tool.get('selections');
          if (!tool.get('isDrawing')) {
            selections.get('notes').forEach((note: MidiNote) => {
              dispatch(removeNote(note));
            });
            selections.get('notes').forEach((note: MidiNote) => {
              dispatch(
                addNote(note.set('startBeat', note.get('startBeat') + 0.25))
              );
            });
            dispatch(setSelections(selections.slideBeat(0.25)));
          }
        }
        break;
      case 'decreaseBeat':
        {
          const selections = tool.get('selections');
          if (!tool.get('isDrawing')) {
            selections.get('notes').forEach((note: MidiNote) => {
              dispatch(removeNote(note));
            });
            selections.get('notes').forEach((note: MidiNote) => {
              dispatch(
                addNote(note.set('startBeat', note.get('startBeat') - 0.25))
              );
            });
            dispatch(setSelections(selections.slideBeat(-0.25)));
          }
        }
        break;
      case 'removeNotes': {
        const selections = tool.get('selections');
        if (!tool.get('isDrawing')) {
          selections.get('notes').forEach((note: MidiNote) => {
            dispatch(removeNote(note));
          });
          dispatch(clearSelections());
        }
        break;
      }
      case 'keepSelectionNotes':
        dispatch(setKeepSelections(true));
        break;
      default:
        return {
          type: 'ERROR',
          payload: {},
          meta: {},
          error: new Error(`${matchedAction} is not defined.`)
        };
    }
  };
}

export function onKeyUp(event) {
  return (dispatch: Dispatch, getState: GetState) => {
    const { editor, preferences } = getState();
    const tool = editor.get('tool');
    const matchedAction = matchedActionName(event, preferences.keyBinds, true);
    switch (matchedAction) {
      case 'keepSelectionNotes':
        dispatch(setKeepSelections(false));
        break;
      default:
        return {
          type: 'ERROR',
          payload: {},
          meta: {},
          error: new Error(`${matchedAction} is not defined.`)
        };
    }
  };
}
