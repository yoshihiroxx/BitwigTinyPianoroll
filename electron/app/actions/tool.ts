import { List } from 'immutable';
import MidiClip from '../models/MidiClip';
import MidiNote from '../models/MidiNote';
import { PenTool, EraserTool, MoveTool, RectTool } from '../tool/Tools';

import StretchTool from '../tool/StretchTool';
import LengthTool from '../tool/LengthTool';
import { GetState, Dispatch } from '../reducers/types';

export const HANDLE_CLICK = 'HANDLE_CLICK';
export const HANDLE_TOOL = 'HANDLE_TOOL';
export const HANDLE_DRAG = 'HANDLE_DRAG';
export const HANDLE_RELEASE = 'HANDLE_RELEASE';
export const HANDLE_EVENT = 'HANDLE_EVENT';
export const HANDLE_MOUSE_EVENT = 'HANDLE_MOUSE_EVENT';
export const ADD_NOTE = 'ADD_NOTE';
export const REMOVE_NOTE = 'REMOVE_NOTE';

export function handleEvent(
  event: string,
  beatOrNote: number | MidiNote,
  noteNumber: number
) {
  return {
    type: HANDLE_EVENT,
    event,
    payload: {
      beatOrNote,
      noteNumber
    }
  };
}

export function handleMouseEvent(
  event: string,
  beatOrNote: number | MidiNote,
  noteNumber?: number
) {
  return {
    type: HANDLE_MOUSE_EVENT,
    payload: {
      event,
      beatOrNote,
      noteNumber
    },
    meta: {
      event
    }
  };
}

export function handleTool(toolType: string) {
  return {
    type: HANDLE_TOOL,
    payload: {},
    meta: {
      toolType
    }
  };
}

function addNote(note: MidiNote) {
  return {
    type: ADD_NOTE,
    payload: {
      note
    },
    meta: {},
    error: null
  };
}

function removeNote(note: MidiNote) {
  return {
    type: REMOVE_NOTE,
    payload: {
      note
    },
    meta: {},
    error: null
  };
}

function handleMouseRelease(
  event: string,
  beatOrNote: number | MidiNote,
  noteNumber: number
) {
  return (dispatch: Dispatch, getState: GetState) => {
    const { editor } = getState();
    const tool = editor.get('tool');

    const selections = tool.get('selections');
    if (tool instanceof EraserTool) {
      selections.get('notes').forEach((note: MidiNote) => {
        dispatch(removeNote(note));
      });
    } else if (tool instanceof MoveTool) {
      selections.get('notes').forEach((note: MidiNote) => {
        dispatch(removeNote(note));
      });
    }

    const drawing = editor.getIn(['tool', 'drawing']);
    if (tool instanceof PenTool) {
      drawing.get('notes').forEach((note: MidiNote) => {
        dispatch(addNote(note));
      });
    } else if (tool instanceof MoveTool) {
      drawing.get('notes').forEach((note: MidiNote) => {
        dispatch(addNote(note));
      });
    }

    dispatch(handleMouseEvent('release', 0, 0));
    dispatch(handleTool('pen'));
  };
}

export function onMouseEvent(
  event: string,
  beatOrNote: number | MidiNote,
  noteNumber?: number
) {
  switch (event) {
    case 'click':
      return handleMouseEvent(event, beatOrNote, noteNumber);
    case 'drag':
      return handleMouseEvent(event, beatOrNote, noteNumber);
    case 'release':
      return handleMouseRelease();
    default:
      return {
        type: 'ERROR',
        payload: {},
        meta: {},
        error: new Error(`${event} is not defined.`)
      };
  }
}

export function focusClip(trackId: number, clipId: number) {
  return {
    type: 'HANDLE_CLIP',
    payload: {
      trackId,
      clipId
    }
  };
}
