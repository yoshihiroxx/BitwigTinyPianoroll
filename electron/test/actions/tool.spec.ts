import MidiNote from '../../app/models/MidiNote';
import { onMouseEvent } from '../../app/actions/tool';
import { ActionType } from '../../app/reducers/types';

describe('actions', () => {
  test('onMouseEvent behavior', () => {
    let action = onMouseEvent('click', 0, 0);
    expect(action).toHaveProperty('type');
    expect(action).toHaveProperty('payload');
    expect(typeof action.payload.beatOrNote).toEqual('number');
    expect(typeof action.payload.noteNumber).toEqual('number');
    expect(action).toHaveProperty('meta');

    action = onMouseEvent('click', new MidiNote());
    expect(action).toHaveProperty('type');
    expect(action).toHaveProperty('payload');
    expect(action.payload.beatOrNote instanceof MidiNote).toBeTruthy();
    expect(action).toHaveProperty('meta');

    action = onMouseEvent('drag', 0, 0);
    expect(action).toHaveProperty('type');
    expect(action).toHaveProperty('payload');
    expect(typeof action.payload.beatOrNote).toEqual('number');
    expect(typeof action.payload.noteNumber).toEqual('number');
    expect(action).toHaveProperty('meta');

    action = onMouseEvent('drag', new MidiNote());
    expect(action).toHaveProperty('type');
    expect(action).toHaveProperty('payload');
    expect(action.payload.beatOrNote instanceof MidiNote).toBeTruthy();
    expect(action).toHaveProperty('meta');

    action = onMouseEvent('release', 0, 0);
    expect(action).toMatchSnapshot();

    action = onMouseEvent('release', new MidiNote());
    expect(action).toMatchSnapshot();
  });
});
