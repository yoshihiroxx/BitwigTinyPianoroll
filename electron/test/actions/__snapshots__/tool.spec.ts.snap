// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions onMouseEvent behavior 1`] = `[Function]`;

exports[`actions onMouseEvent behavior 2`] = `[Function]`;
